diff -ur volatility/volatility/conf.py volatility_py/volatility/conf.py
--- volatility/volatility/conf.py	2016-12-27 10:54:47.000000000 -0500
+++ volatility_py/volatility/conf.py	2017-06-22 06:11:31.906056022 -0400
@@ -89,7 +89,8 @@
         ## We cant emit errors about missing parameters until we are
         ## sure that all modules have registered all their parameters
         if self.final:
-            return optparse.OptionParser.error(self, msg)
+            raise Exception(msg)
+            #return optparse.OptionParser.error(self, msg)
         else:
             raise RuntimeError(msg)
 
@@ -260,7 +261,7 @@
                             pass
 
                     self.optparser.print_help()
-                    sys.exit(0)
+                    return False
             except AttributeError:
                 pass
 
@@ -268,6 +269,7 @@
             import volatility.cache as cache
             for k, v in self.cache_invalidators.items():
                 cache.CACHE.invalidate_on(k, v)
+        return True
 
     def remove_option(self, option):
         """ Removes options both from the config file parser and the
