diff --git a/volatility/conf.py b/volatility/conf.py
index 6c65d72..f29b663 100755
--- a/volatility/conf.py
+++ b/volatility/conf.py
@@ -89,7 +89,8 @@ class PyFlagOptionParser(optparse.OptionParser):
         ## We cant emit errors about missing parameters until we are
         ## sure that all modules have registered all their parameters
         if self.final:
-            return optparse.OptionParser.error(self, msg)
+            raise Exception(msg)
+            #return optparse.OptionParser.error(self, msg)
         else:
             raise RuntimeError(msg)
 
@@ -99,6 +100,10 @@ class PyFlagOptionParser(optparse.OptionParser):
         for cb in self.help_hooks:
             file.write(cb())
 
+#Just add this member so that we can check if we are importing the
+#correct volatility installation
+PyREBoxVolatility = True
+
 class ConfObject(object):
     """ This is a singleton class to manage the configuration.
 
@@ -260,7 +265,7 @@ class ConfObject(object):
                             pass
 
                     self.optparser.print_help()
-                    sys.exit(0)
+                    return False
             except AttributeError:
                 pass
 
@@ -268,6 +273,7 @@ class ConfObject(object):
             import volatility.cache as cache
             for k, v in self.cache_invalidators.items():
                 cache.CACHE.invalidate_on(k, v)
+        return True
 
     def remove_option(self, option):
         """ Removes options both from the config file parser and the
